engine/execution/ingestion/stop_control.go

```diff
@@ -3,318 +3,567 @@ package ingestion
+		header, err := s.headers.ByHeight(s.stopBoundary.StopHeight - 1)
 		if err != nil {
-			// any error here would indicate unexpected storage error, so we crash the node
-			// TODO: what if the error is due to the node being stopped?
-			// i.e. context cancelled?
-			s.log.Fatal().
-				Err(err).
-				Str("block_id", h.ID().String()).
-				Msg("failed to check if the block has been executed")
+			handleErr(fmt.Errorf("failed to get header by height: %w", err))
 			return
 		}
+		parentID = header.ID()
+	}
 
-		if executed {
-			s.stopExecution()
-		} else {
-			s.stopAfterExecuting = h.ParentID
-			s.log.Info().
-				Msgf(
-					"Node scheduled to stop executing"+
-						" after executing block %s at height %d",
-					s.stopAfterExecuting.String(),
-					h.Height-1,
-				)
-		}
+	s.stopBoundary.stopAfterExecuting = parentID
+
+	s.log.Info().
+		Stringer("block_id", h.ID()).
+		Stringer("stop", s.stopBoundary).
+		Stringer("stop_after_executing", s.stopBoundary.stopAfterExecuting).
+		Msgf("Found ID of the block that should be executed last")
+
+	// check if the parent block has been executed then stop right away
+	executed, err := state.IsBlockExecuted(ctx, execState, h.ParentID)
+	if err != nil {
+		handleErr(fmt.Errorf(
+			"failed to check if the block has been executed: %w",
+			err,
+		))
+		return
+	}
+
+	if executed {
+		// we already reached the point where we should stop
+		s.stopExecution()
+		return
 	}
 }
 
-// blockExecuted should be called after a block has finished execution
-func (s *StopControl) blockExecuted(h *flow.Header) {
+// OnBlockExecuted should be called after a block has finished execution
+func (s *StopControl) OnBlockExecuted(h *flow.Header) {
 	s.Lock()
 	defer s.Unlock()
 
-	if s.state == StopControlPaused || s.state == StopControlOff {
+	if s.stopBoundary == nil || s.stopped {
 		return
 	}
 
-	if s.stopAfterExecuting == h.ID() {
-		// double check. Even if requested stopHeight has been changed multiple times,
-		// as long as it matches this block we are safe to terminate
-		if h.Height == s.stopHeight-1 {
-			s.stopExecution()
-		} else {
-			s.log.Warn().
-				Msgf(
-					"Inconsistent stopping state. "+
-						"Scheduled to stop after executing block ID %s and height %d, "+
-						"but this block has a height %d. ",
-					h.ID().String(),
-					s.stopHeight-1,
-					h.Height,
-				)
-		}
+	if s.stopBoundary.stopAfterExecuting != h.ID() {
+		return
+	}
+
+	// double check. Even if requested stopHeight has been changed multiple times,
+	// as long as it matches this block we are safe to terminate
+	if h.Height != s.stopBoundary.StopHeight-1 {
+		s.log.Warn().
+			Msgf(
+				"Inconsistent stopping state. "+
+					"Scheduled to stop after executing block ID %s and height %d, "+
+					"but this block has a height %d. ",
+				h.ID().String(),
+				s.stopBoundary.StopHeight-1,
+				h.Height,
+			)
+		return
 	}
+
+	s.stopExecution()
 }
 
 func (s *StopControl) stopExecution() {
```

Comment:
```suggestion
// Caller must acquire the lock.
func (s *StopControl) stopExecution() {
```
would suggest to add this to all the private unsynchronized functions.

Reply:
[b61af2a](https://github.com/onflow/flow-go/pull/4342/commits/b61af2a8773709c672ee17ccd6e777b08aa20daa)